sys/signal.c

```diff
@@ -0,0 +1,221 @@
+#include <signal.h>
+#include <thread.h>
+#include <malloc.h>
+#include <stdc.h>
+#include <errno.h>
+#include <sysent.h>
+#include <systm.h>
+
+static MALLOC_DEFINE(sig_pool, "signal handlers pool");
+
+static sighandler_t *signal_default_actions[SIG_LAST] = {
+    [SIGINT] = SIG_TERM,  [SIGILL] = SIG_TERM,  [SIGFPE] = SIG_TERM,
+    [SIGABRT] = SIG_TERM, [SIGSEGV] = SIG_TERM, [SIGKILL] = SIG_TERM,
+    [SIGCHLD] = SIG_IGN,  [SIGUSR1] = SIG_TERM, [SIGUSR2] = SIG_TERM,
+};
+
+static const char *signal_names[SIG_LAST] = {
+    [SIGINT] = "SIGINT",   [SIGILL] = "SIGILL",   [SIGABRT] = "SIGABRT",
+    [SIGFPE] = "SIGFPE",   [SIGSEGV] = "SIGSEGV", [SIGKILL] = "SIGKILL",
+    [SIGTERM] = "SIGTERM", [SIGCHLD] = "SIGCHLD", [SIGUSR1] = "SIGUSR1",
+    [SIGUSR2] = "SIGUSR2",
+};
+
+void sighand_init() {
+  /* TODO: A lot of kernel components have a init function whose only purpose is
+     to perform a kmalloc_init. Maybe we could collect them with a linker set,
+     and initialize all pools together? */
+  kmalloc_init(sig_pool, 2, 2);
+}
+
+sighand_t *sighand_new() {
+  sighand_t *sh = kmalloc(sig_pool, sizeof(sighand_t), M_ZERO);
```

Comment:
[APUE](https://en.wikipedia.org/wiki/Advanced_Programming_in_the_Unix_Environment) §12.8 explains explicitely that all threads within given process share same `sighand_t` structure:

> Each thread has its own signal mask, but the signal disposition is shared by all
threads in the process. As a consequence, individual threads can block signals, but
when a thread modifies the action associated with a given signal, all threads share the
action. Thus, if one thread chooses to ignore a given signal, another thread can undo
that choice by restoring the default disposition or installing a signal handler for that
signal.

So that suggests `sighand_t` should be a member of process structure. The problem is we don't have one. Should we finally introduce the distinction? Let's discuss it during the meeting.